### PR TITLE
Added Device/Collection/Distribution Metadata API

### PR DESCRIPTION
**Pull request by kk25q**
Ready for Review.

### CODE REVIEW

File: src/Collection.php [hunk 1]
```diff
@@ -67,24 +67,54 @@ public function removeDevice($device_id) {
...
+ * @param array $params
+ * @return HttpResponse
+ */
+  public function metadataField($key, $params = array()) {
+    return $this->client->get($this->path() . '/metadata/' . $key, $params);
+  }
+
+/**
+ * Update Collection Metadata Field.
+ *
```
Comment (sarahhenkens) [inline src/Collection.php]: There should only be one empty row between the function description and the `@link`.

File: src/Device.php [hunk 2]
```diff
@@ -296,4 +296,60 @@ public function process($command, $data = null) {
...
+ * @return HttpResponse
+ */
+  public function metadataField($key, $params = array()) {
+    return $this->client->get($this->path() . '/metadata/' . $key, $params);
+  }
+
+/**
+ * Update Device Metadata.
+ *
+ *
```
Comment (sarahhenkens) [inline src/Device.php]: There should only be one empty row between the function description and the `@link`.

File: src/Device.php [hunk 3]
```diff
@@ -296,4 +296,60 @@ public function process($command, $data = null) {
...
+ * @param array $params
+ * @return HttpResponse
+ */
+  public function updateMetadata($params = array()) {
+    return $this->client->put($this->path() . '/metadata', $params);
+  }
+
+/**
+ * Update Device Metadata Field.
+ *
```
Comment (sarahhenkens) [inline src/Device.php]: There should only be one empty row between the function description and the `@link`.

File: src/Distribution.php [hunk 4]
```diff
@@ -85,4 +85,61 @@ public function stream($name) {
...
+ * @param array $params
+ * @return HttpResponse
+ */
+  public function metadataField($key, $params = array()) {
+    return $this->client->get($this->path() . '/metadata/' . $key, $params);
+  }
+
+/**
+ *  Update metadata of a distribution device.
+ *
```
Comment (sarahhenkens) [inline src/Distribution.php]: There should only be one empty row between the function description and the `@link`.

File: src/Distribution.php [hunk 5]
```diff
@@ -85,4 +85,61 @@ public function stream($name) {
...
+ * @param array $params
+ * @return HttpResponse
+ */
+  public function updateMetadata($params = array()) {
+    return $this->client->put($this->path() . '/metadata', $params);
+  }
+
+/**
+ * Update distribution device metadata field.
+ *
```
Comment (sarahhenkens) [inline src/Distribution.php]: There should only be one empty row between the function description and the `@link`.

File: src/Collection.php [hunk 6]
```diff
@@ -67,24 +67,54 @@ public function removeDevice($device_id) {
...
  *
- * @return array
+ * @param array $params
+ * @return HttpResponse
  */
-  public function metadata() {
-    $response = $this->client->get($this->path() . '/metadata');
-    return $response->json();
+  public function metadata($params = array()) {
+    return  $this->client->get($this->path() . '/metadata', $params);
```
Comment (sarahhenkens) [inline src/Collection.php]: There are 2 spaces between `return  $this`. Should only be one.

File: examples/device_metadata.php [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      echo $response->raw();
+    }
+
+    $metadataObj = array("hi"=> "distribution");
+    $response = $distribution->updateMetadata($metadataObj);
+    if ($response->statusCode == 204) {
+       echo  "\n\rUpdate metadata for distribution is Successful.";
+       $response = $distribution->updateMetadataField("hi", "distributions");
+       if ($response->statusCode == 204) {
+            echo  "\n\rUpdate metadata field for distribution is Successful.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: Too many spaces between `echo` and the string. This applies to various lines in this file.

File: src/Device.php [hunk 8]
```diff
@@ -296,4 +296,60 @@ public function process($command, $data = null) {
...
+ *
+ * @link https://m2x.att.com/developer/documentation/v2/device#Update-Device-Metadata-Field
+ *
+ * @param string $key
+ * @param string $value
+ * @return HttpResponse
+ */
+  public function updateMetadataField($key, $value) {
+      return $this->client->put($this->path() . '/metadata/' . $key , array( value=> $value));
```
Comment (sarahhenkens) [inline src/Device.php]: This should be `array('value' => $value)`.

File: src/Distribution.php [hunk 9]
```diff
@@ -85,4 +85,61 @@ public function stream($name) {
...
+ *
+ * @link https://m2x.att.com/developer/documentation/v2/distribution#Update-Distribution-Metadata-Field
+ *
+ * @param string $key
+ * @param string $value
+ * @return HttpResponse
+ */
+  public function updateMetadataField($key, $value) {
+      return $this->client->put($this->path() . '/metadata/' . $key , array( value=> $value));
```
Comment (sarahhenkens) [inline src/Distribution.php]: This should be `array('value' => $value)`.

**Comment by kk25q** [comment]
@sarahhenkens Requested Changes are Completed.Ready For Review.

File: examples/device_metadata.php [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    echo "\n\rLoad metadata for device ";
+    echo $response->raw();
+    $metadata = json_decode($response->raw(), true);
+    if (count($metadata) > 0) {
+      $response = $device->metadataField($metadata[0]['key']);
+      echo "\n\rLoad metadata field for device ";
+      echo $response->raw();
+    }
+
+    $metadataObj = array("hi"=> "hello");
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: There should be a space around `=>`.

Best-practice example:

```php
array("foo" => "bar");
```

File: examples/device_metadata.php [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    $response = $device->updateMetadata($metadataObj);
+    if ($response->statusCode == 204) {
+       echo "\n\rUpdate metadata for device is Successful.";
+       $response = $device->updateMetadataField("hi", "test");
+       if ($response->statusCode == 204) {
+            echo "\n\rUpdate metadata field for device is Successful.";
+       } else {
+            echo "\n\rUpdate metadata field for device Failed.Please Try Again.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: Missing space after the `.` in the first sentence. "Please try again" should only have its first letter capitalized.

File: examples/device_metadata.php [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+       }
+    } else {
+       echo "\n\rUpdate metadata for device Failed.Please Try Again.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: Missing space after the `.` in the first sentence.

File: examples/device_metadata.php [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+      echo "\n\rLoad metadata field for collection ";
+      echo $response->raw();
+    }
+    $metadataObj = array("hi"=> "hello");
+    $response = $collection->updateMetadata($metadataObj);
+    if ($response->statusCode == 204) {
+       echo "\n\rUpdate metadata for collection is Successful.";
+       $response = $collection->updateMetadataField("hi", "Test");
+       if ($response->statusCode == 204) {
+            echo "\n\rUpdate metadata field for collection is Successful.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: "successful" should not be capitalized.

File: examples/device_metadata.php [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+       } else {
+            echo "\n\rUpdate metadata field for collection Failed.Please Try Again.";
+       }
+    } else {
+       echo "\n\rUpdate metadata for collection Failed.Please Try Again.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: Missing space after the `.` in the first sentence. "Please try again" should only have its first letter capitalized.

File: examples/device_metadata.php [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    $metadataObj = array("hi"=> "hello");
+    $response = $collection->updateMetadata($metadataObj);
+    if ($response->statusCode == 204) {
+       echo "\n\rUpdate metadata for collection is Successful.";
+       $response = $collection->updateMetadataField("hi", "Test");
+       if ($response->statusCode == 204) {
+            echo "\n\rUpdate metadata field for collection is Successful.";
+       } else {
+            echo "\n\rUpdate metadata field for collection Failed.Please Try Again.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: Missing space after the `.` in the first sentence. "Please try again" should only have its first letter capitalized.

File: examples/device_metadata.php [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    if (count($metadata) > 0) {
+      $response = $distribution->metadataField($metadata[0]['key']);
+      echo "\n\rLoad metadata field for distribution ";
+      echo $response->raw();
+    }
+
+    $metadataObj = array("hi"=> "distribution");
+    $response = $distribution->updateMetadata($metadataObj);
+    if ($response->statusCode == 204) {
+       echo "\n\rUpdate metadata for distribution is Successful.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: "Successful" should not be capitalized.

File: examples/device_metadata.php [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+       $response = $distribution->updateMetadataField("hi", "distributions");
+       if ($response->statusCode == 204) {
+            echo "\n\rUpdate metadata field for distribution is Successful.";
+       } else {
+            echo "\n\rUpdate metadata field for distribution Failed.Please Try Again.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: Missing space after the `.` in the first sentence. "Please try again" should only have its first letter capitalized.

File: examples/device_metadata.php [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+      echo $response->raw();
+    }
+
+    $metadataObj = array("hi"=> "distribution");
+    $response = $distribution->updateMetadata($metadataObj);
+    if ($response->statusCode == 204) {
+       echo "\n\rUpdate metadata for distribution is Successful.";
+       $response = $distribution->updateMetadataField("hi", "distributions");
+       if ($response->statusCode == 204) {
+            echo "\n\rUpdate metadata field for distribution is Successful.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: "successful" should not be capitalized.

File: examples/device_metadata.php [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+       } else {
+            echo "\n\rUpdate metadata field for distribution Failed.Please Try Again.";
+       }
+    } else {
+       echo "\n\rUpdate metadata for distribution Failed.Please Try Again.";
```
Comment (sarahhenkens) [inline examples/device_metadata.php]: Missing space after the `.` in the first sentence. "Please try again" should only have its first letter capitalized.

File: src/Collection.php [hunk 20]
```diff
@@ -67,24 +67,53 @@ public function removeDevice($device_id) {
...
   }
+
+/**
+ * Get Collection Metadata Field.
+ *
+ * Get the value of a single custom metadata field from an existing Collection.
+ *
+ * @link https://m2x.att.com/developer/documentation/v2/collections#Read-Collection-Metadata-Field
+ *
+ * @param array $key
```
Comment (sarahhenkens) [inline src/Collection.php]: The `$key` param is a `string` type.

File: src/Collection.php [hunk 21]
```diff
@@ -67,24 +67,53 @@ public function removeDevice($device_id) {
...
+ * Update Collection Metadata Field.
+ *
+ * @link https://m2x.att.com/developer/documentation/v2/collections#Update-Collection-Metadata-Field
+ *
+ * @param string $key
+ * @param string $value
+ * @return HttpResponse
+ */
+  public function updateMetadataField($key, $value) {
+      return $this->client->put($this->path() . '/metadata/' . $key , array( value=> $value));
```
Comment (sarahhenkens) [inline src/Collection.php]: `array( value=> $value))` should be `array('value' => $value)`. Make sure to test your code.

**Comment by sarahhenkens** [comment]
Make sure to squash commits before merging.

**Comment by kk25q** [comment]
@sarahhenkens latest commits are squashed.Ready for Review.

**Comment by sarahhenkens** [comment]
Ready for merge

**Pull request merged**